openfl/utilities/optimizers/numpy/adagrad_optimizer.py

```diff
@@ -0,0 +1,86 @@
+
+import numpy as np
+
+from .base_optimizer import Optimizer
+
+
+class NumpyAdagrad(Optimizer):
+    """Adagrad optimizer implementation.
+
+    Original paper: http://jmlr.org/papers/v12/duchi11a.html
+    """
+
+    def __init__(
+        self,
+        params: Optional[Dict[str, np.ndarray]] = None,
+        *,
+        model_interface=None,
+        learning_rate: float = 0.01,
+        initial_accumulator_value: float = 0.1,
+        epsilon: float = 1e-10,
+    ) -> None:
+        """Initialize.
+
+        Args:
+            params: Parameters to be stored for optimization.
+            model_interface: Model interface instance to provide parameters.
+            learning_rate: Tuning parameter that determines
+                the step size at each iteration.
+            initial_accumulator_value: Initial value for squared gradients.
+            epsilon: Value for computational stability.
+        """
+        super().__init__()
+
+        assert model_interface is not None or params, (
+            'Should provide one of the params or model_interface')
+
+        if learning_rate < 0:
+            raise ValueError(
+                f'Invalid learning rate: {learning_rate}. Learning rate must be >= 0.')
+        if initial_accumulator_value < 0:
+            raise ValueError(
+                f'Invalid initial_accumulator_value value: {initial_accumulator_value}. \
+                Initial accumulator value must be >= 0.')
+        if epsilon <= 0:
+            raise ValueError(
+                f'Invalid epsilon value: {epsilon}. Epsilon avalue must be > 0.')
+
+        self.params = params
+
+        if params is None and model_interface is not None:
+            self._set_params_from_model(model_interface)
+
+        self.learning_rate = learning_rate
+        self.initial_accumulator_value = initial_accumulator_value
+        self.epsilon = epsilon
+
+        self.grads_squared = {}
+        for param_name in self.params:
+            self.grads_squared[param_name] = np.full_like(self.params[param_name],
+                                                          self.initial_accumulator_value)
+
+    def _update_param(self, grad_name: str, grad: np.ndarray) -> None:
+        """Update papams by given gradients."""
+        self.params[grad_name] = (self.params[grad_name] - self.learning_rate * grad
+                                  / (np.sqrt(self.grads_squared[grad_name]) + self.epsilon))
+
+    def step(self, gradients: Dict[str, np.ndarray]) -> None:
+        """
+        Perform a single step for parameter update.
+
+        Implement Adagrad optimizer weights update rule.
+
+        Args:
+            gradients: Partial derivatives with respect to optimized parameters.
+        """
+        for grad_name in gradients:
+            if grad_name not in self.grads_squared:
+                raise KeyError(f"Key {grad_name} doesn't exist in optimized parameters")
+
+            grad = gradients[grad_name]
```

Comment:
`numpy-ml` having such code

```python
        # scale gradient to avoid explosion
        t = np.inf if clip_norm is None else clip_norm
        if norm(param_grad) > t:
            param_grad = param_grad * t / norm(param_grad)
```
does it have sense?

Reply:
@dmitryagapov by default `clip_norm` is None - 
https://github.com/ddbourgin/numpy-ml/blob/b0359af5285fbf9699d64fd5ec059493228af03e/numpy_ml/neural_nets/optimizers/optimizers.py#L60
I can add this parameter, but I'm not sure if it's really needed.